test-tests/shared/src/test/scala/zio/test/SpySpec.scala

```diff
@@ -0,0 +1,76 @@
+package zio.test
+
+import zio.test.Assertion._
+import zio.test.mock._
+import zio.{ Has, Ref, UIO, ZIO, ZLayer }
+
+object SpySpec extends DefaultRunnableSpec {
+
+  type Counter = Has[Counter.Service]
+
+  object Counter {
+
+    trait Service {
+      def increment: UIO[Unit]
+      def decrement: UIO[Unit]
+      def get: UIO[Int]
+      def reset: UIO[Unit]
+    }
+
+    final case class Live(counterState: Ref[Int]) extends Service {
+      def increment: UIO[Unit] = counterState.update(_ + 1).unit
+      def decrement: UIO[Unit] = counterState.update(_ - 1).unit
+      def get: UIO[Int]        = counterState.get
+      def reset: UIO[Unit]     = counterState.set(0)
+    }
+
+    val live: ZLayer.NoDeps[Nothing, Counter] =
+      ZLayer.fromEffect(Ref.make(0).map(ref => Has(Live(ref))))
+  }
+
+  sealed trait Command[+A] extends Method[Counter.Service, Unit, A]
+
+  object Command {
+    case object Increment extends Command[Unit]
+    case object Decrement extends Command[Unit]
+    case object Get       extends Command[Int]
+    case object Reset     extends Command[Unit]
+
+    val commands = Gen.elements(Increment, Decrement, Get, Reset)
+  }
+
+  val spy: ZLayer[Counter, Nothing, Counter with Spy] =
+    Spy.live[Int, Counter.Service](0) { (counter, state, result) =>
+      new Counter.Service {
+        def increment: UIO[Unit] =
+          counter.increment *> state.update(_ + 1).unit
```

Comment:
There is a bit of boilerplate here. The ideal way to do this might be something like:

```scala
PartialFunction[(Ref[State], Method[A, B], A, B), UIO[Unit]]
```

That way you get to see the method, the input, the output, and then modify the state as per appropriate to your spy.

Now this can't exactly be written today. We have `Mockable` to go from `Mock => A`, but we don't have something like `Unmockable` (`Spyable`) to go the other way: to deconstruct one or more calls on `A` into the individual methods, with inputs and outputs. But if we had such a thing and if the macros project could generate them automatically, then we might be able to eliminate all the boilerplate, and make it easy to create many different spies for the same interface.

Reply:
I revised along the lines you suggested. There is still some boilerplate in implementing the `Spyable` interface but now you only have to do it once for your service and can create many different spies based on that. Also agree it would be very amenable to macros.

I ran into some issues with the input `I` parameter in `Method` being invariant and realized we could make it contravariant by moving a couple of the methods that make assertions based on the input to an implicit syntax class. I think that makes more sense as it reflects a `Method[M, I, A]` modeling a `Function[I, A]`. I think we may also want to consider changing the way we model functions that don't take any inputs from `Method[M, Unit, A]` to `Method[M, Any, A]`. A method that doesn't require any inputs can be run with any input at all and while the unit value is thus one value that would be sufficient there is nothing special about that. This is also more consistent with our treatment of the environment and seems to work better when we are recovering type information by pattern matching against methods.

Reply:
I agree with you. We should make it `Any` and not `Unit`. 💯 💪